docs/dag-sharding-rfc.md

```diff
@@ -0,0 +1,176 @@
+
+If we are in a situation where shards approach the size of ipfs repo or working memory then we can gather inspiration from the report by Plank, specifically the section "Implementation and Performance Details: Checkpointing Systems".  In this section Plank outlines two algorithms for setting checksum values after the data shards are already stored by sending messages over the network.  From my first read-through the broadcast algorithm looks the most promising.  This algorithm would allow cluster to send shards one at a time to the peer holding a zeroed out checksum shard and then perform successive updates to calculate the checksums, rather than requiring that the cluster-dag importer hold the one shard being filled up for pinning alongside the m checksum shards being calculated.
+
+
+### Automatic Recovery
+TODO -- I have thoughts about implementing this in cluster that I need to organize and put down.
+
+# Complications
+
+## Files larger than local repos
+
+### Improve cluster UX for large files
+"How much room do I have in my cluster?" is a very predictable question that users importing huge files will want to know.  We should include cluster cli tools (something like ipfs metrics, or the still-unmade connectivity-graph) that inform users how much space they have available, and the largest recommended file the cluster can ingest.  Perhaps later on we can get even fancier and include predictions of insertion time, and a visualization of how many shards of what size will be added.
+
+Although the ipfs add proxy endpoint is the logical place to launch the adding process we could also provide a user friendly command that takes in the option of a network address to add from.  This way cluster could properly manage the connection with a remote server while buffering and potentially blocking for long periods of time and prevent the user from having to do clever unix things to download large files.  Note I don't have much experience with this kind of thing, so maybe wget etc have really good UX for things like this and this would be overkill.
+
+### Challenges adding files
+When files become larger than local repos this becomes challenging for two reasons.  First, the importing node now needs to be able to block the importing process periodically to transfer data over to the pinning peer so that its local resources (ipfs repo and maybe memory) don't get overloaded.  Many related complications also seem likely.  For example if one shard takes up 90% of the node's ipfs repo and if the importing node makes use of its repo for temporary pinning and transfer then the importing node cannot pin its own shard until all other nodes have been served.  Second, I am unclear if either of the two proposed data transfer methods will provide acceptable performance when transferring large amounts of data (e.g. on the order of 10 GB shards).
```

Comment:
When cluster ingests data, it should chunk and forward each chunk to the final destination. This should not take local disk space and should be memory-conservative. Perhaps some intelligent chunking methods require to read a lot of data first (in a way that you have a ram memory bound), but we'll have to live with that. I'm thinking under the assumption that data can be chunked as it comes in in most cases (normally format-indexes are in headers so reading those headers helps you chunk the rest as it comes in).

Reply:
Thanks totally addresses concerns here